csv2bufr/__init__.py

```diff
@@ -196,258 +168,303 @@ def validate_value(key: str, value: Union[NUMBERS],
+            elif "eccodes_key" in object_:
+                result = self.get_element( object_["eccodes_key"] )
+                if result in (CODES_MISSING_LONG, CODES_MISSING_DOUBLE):
+                    result = None
+            else:
+                for k in object_:
+                    object_[k] = self._extract(object_[k])
+                result = object_
+        elif isinstance(object_, list):
+            for idx in range(len(object_)):
+                object_[idx] = self._extract(object_[idx])
             result = object_
-    elif isinstance(object_, list):
-        for idx in range(len(object_)):
-            object_[idx] = extract(bufr_msg, object_[idx])
-        result = object_
-    else:
-        result = object_
-
-    return result
-
-
-def transform(data: str, mappings: dict, station_metadata: dict) -> dict:
-    """
-    TODO: describe function
-
-    :param data: TODO: describe
-    :param mappings: TODO: describe
-    :param station_metadata: TODO: describe
-
-    :return: `dict` of BUFR messages
-    """
-
-    # validate mappings
-    e = validate_mapping_dict(mappings)
-    if e is not SUCCESS:
-        raise ValueError("Invalid mappings")
-
-    LOGGER.debug("mapping dictionary validated")
-
-    # TODO: add in code to validate station_metadata
-
-    # we may have multiple rows in the file, create list object to return
-    # one item per message
-    messages = {}
-    # now convert data to StringIO object
-    fh = StringIO(data)
-    # now read csv data and iterate over rows
-    reader = csv.reader(fh, delimiter=',', quoting=csv.QUOTE_NONNUMERIC)
-    rows_read = 0
-    for row in reader:
-        if rows_read == 0:
-            col_names = row
         else:
-            data = row
-            data_dict = dict(zip(col_names, data))
-            try:
-                data_dict = {**data_dict, **station_metadata['data']}
-            except Exception as e:
-                message = "issue merging station and data dictionaries."
-                LOGGER.error(f"{message}{e}")
-                raise e
-            # Iterate over items to map, perform unit conversions and validate
-            for element in mappings["sequence"]:
-                value = element["value"]
-                column = element["column"]
-                # select between "value" and "column" fields.
-                if value is not None:
-                    value = element["value"]
-                elif column is not None:
-                    # get column name
-                    # make sure column is in data_dict
-                    if (column not in data_dict):
-                        message = f"column '{column}' not found in data dictionary"  # noqa
-                        raise ValueError(message)
-                    value = data_dict[column]
-                    if value in MISSING:
+            result = object_
+        return result
+
+    def parse(self, data: str, metadata: dict, mappings: dict):
```

Comment:
add return type in function annotation, add docstring

Reply:
done

Reply:
done